### PR TITLE
Autotest against llvm-dwarfdump

### PR DESCRIPTION
**Pull request by sevaa**
This is the first draft of a script that will mimic the output of llvm-dwarfdump. The original has a brief mode and a verbose mode; we only mimic the latter.

The only option tested (section dumped) is debug-info. It's more involved that readelf's conterpart; for one thing, it dumps the referenced loclists and rangelists, so test coverage will include parsing those, and expressions in those. More importantly, the output of auxiliary section dumps conceptually doesn't match between readelf and dwarfdump (they iterate over different things), the latter fails on a binary by GCC8.

I've included an autotest harness similar to what the readelf autotest does. There are three files so far, ones with recent, involved DWARF. What are other DWARF examples that we have and that we could test it on?

One big challenge was - **dwarfdump spells out the datatypes of variables and parameters**. I had a rather sloppy script for that, I've included that in the library. It's a work in progress, as C++' datatype system is hairy. As more binaries come in, more work will surely be needed. If the coding style of that file is not up to the standards, let me know.

The only nod to dwarfdump's shortcomings was - they don't support all the DWARFv5 operations ( https://github.com/llvm/llvm-project/issues/56365 ), however, there is a failure mode that was easy enough to reproduce and will be easy enough to remove once LLVM catches up.

### CODE REVIEW

File: elftools/dwarf/datatype_cpp.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    if not 'DW_AT_type' in var_die.attributes:
+        t.tag = ''
+        return t
+
+    type_die = var_die.get_DIE_from_attribute('DW_AT_type')
+
+    mods = []
+    # Unlike readelf, dwarfdump doesn't chase typedefs
+    while type_die.tag in ('DW_TAG_const_type', 'DW_TAG_pointer_type', 'DW_TAG_reference_type'):
+        mods.insert(0, type_die.tag[7:-5])
```
[the remark this answers is not in the document]
Comment (sevaa) [reply elftools/dwarf/datatype_cpp.py:44]: [7:-5] makes "foo" from  "DW_TAG_foo_type".

File: test/utils.py [hunk 2]
```diff
@@ -32,13 +32,13 @@ def is_in_rootdir():
     return os.path.isdir('test') and os.path.isdir('elftools')
 
 
-def dump_output_to_temp_files(testlog, *args):
+def dump_output_to_temp_files(testlog, filename, option, *args):
```
[the remark this answers is not in the document]
Comment (sevaa) [reply test/utils.py:35]: Following up from local autotest fails is that much easier when you see which test generated which pair of test files.

File: elftools/dwarf/datatype_cpp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sevaa) [reply elftools/dwarf/datatype_cpp.py:44]: [7:-5] is used in three places, but on different objects. I'll make it more descriptive all the same.

**Comment by eliben** [comment]
When you address the review fully, please ping me in a comment, or click the "Re-request review" button in the list of reviewers next to my name.

Otherwise it's difficult to know when the review is addressed (often people update things in multiple steps with several commits, and so on)

**Comment by sevaa** [comment]
Sorry. I think I've addressed it fully, but you tell me.

**Review by eliben — APPROVED** [review]

**Comment by sevaa** [comment]
Okay. Next up: fix for #424, unless that binary contains indirect forms. Then support for those too. The code is pretty much written.

**Pull request merged**